internal/api/core/response.go

```diff
@@ -70,3 +71,56 @@ func getRequestInfo(request ValidAPIRequest) (string, timestamp.Timestamp) {
 
 	return id, startTime
 }
+
+// Get the endpoint, userEmail, courseID, assignmentID, and locator
+// from a ValidAPIRequest and an APIError, both of which may be nil.
+func getRequestInfoForStats(request ValidAPIRequest, apiError *APIError) (string, string, string, string, string) {
+	endpoint, userEmail, courseID, assignmentID := getValidAPIRequestInfoForStats(request)
+	locator := ""
+
+	if apiError != nil {
+		endpoint = util.GetStringWithDefault(endpoint, apiError.Endpoint)
+		courseID = util.GetStringWithDefault(courseID, apiError.CourseID)
+		assignmentID = util.GetStringWithDefault(assignmentID, apiError.AssignmentID)
+		userEmail = util.GetStringWithDefault(userEmail, apiError.UserEmail)
+		locator = apiError.Locator
+	}
+
+	return endpoint, userEmail, courseID, assignmentID, locator
+}
+
+// Reflexively get the endpoint, userEmail, courseID, and assignmentID from a ValidAPIRequest.
+func getValidAPIRequestInfoForStats(request ValidAPIRequest) (string, string, string, string) {
```

Comment:
I don't like this name.
The "valid" part is tough.
Although it is in the type name of the argument, it is also very overloaded.

Maybe something like: "getBasicAPIRequestInfo".